onnxruntime/contrib_ops/cpu/transformers/beam_search.cc

```diff
@@ -0,0 +1,650 @@
+
+  const T* logits_data = logits.Get<Tensor>().Data<T>();
+
+  // Logits has shape (batch_size * num_beams, input_length, vocab_size),
+  // where input_length equals to parameters_->sequence_length for first subgraph call, and 1 for the remaining calls.
+  const TensorShape& logits_shape = logits.Get<Tensor>().Shape();
+  ORT_ENFORCE(logits_shape.NumDimensions() == 3);
+  auto input_length = logits_shape[1];
+
+  // Get logits for the last token:
+  //    next_token_logits = logits[:, -1, :], and the result shape is (batch_size * num_beams, vocab_size)
+  // When input_length == 1, use logits directly in SoftmaxCPU below so it only need for input_length > 1.
+  gsl::span<T>& next_token_logits = beam_state.next_token_logits;
+  if (input_length > 1) {
+    const T* current_logits = logits_data + (input_length - 1) * vocab_size;
+    for (int i = 0; i < batch_beam_size; i++) {
+      gsl::span<const T> source(current_logits, vocab_size);
+      gsl::span<T> target = next_token_logits.subspan(i * vocab_size, vocab_size);
+      gsl::copy(source, target);
+      current_logits += input_length * vocab_size;
+    }
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  //DumpOrtValue("logits", logits);
+  DumpTensor("next_token_logits", next_token_logits.data(), parameters_->batch_size, parameters_->num_beams, vocab_size);
+#endif
+
+  // Get scores for candidates of next token: next_token_scores = log_softmax(next_token_logits, dim=-1)
+  gsl::span<T>& next_token_scores = beam_state.next_token_scores;
+  Status status = SoftmaxCPU<T>(batch_beam_size,  // rows
+                                vocab_size,       // elements per row
+                                input_length > 1 ? next_token_logits.data() : logits_data,
+                                next_token_scores.data(),
+                                true,
+                                thread_pool_);
+  if (!status.IsOK()) {
+    return status;
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpTensor("next_token_scores after softmax", next_token_scores.data(), parameters_->batch_size, parameters_->num_beams, vocab_size);
+#endif
+
+  // Apply all score processors that updates scores
+  logits_processors_.Process(&(beam_state.sequences), next_token_scores);
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpTensor("next_token_scores after logits processor", next_token_scores.data(), parameters_->batch_size, parameters_->num_beams, vocab_size);
+#endif
+
+  // Add beam score to next token scores. Corresponding python code is like:
+  //    next_token_scores = next_token_scores + beam_scores[:, None].expand_as(next_token_scores)
+  // TODO: use thread pool to parrellel
+  int offset = 0;
+  int batch_beam_index = 0;
+  for (int i = 0; i < parameters_->batch_size; i++) {
+    for (int j = 0; j < parameters_->num_beams; j++, batch_beam_index++) {
+      for (int k = 0; k < parameters_->vocab_size; k++, offset++) {
+        next_token_scores[offset] += beam_state.beam_scores[batch_beam_index];
```

Comment:
Here, I understand we are updating the scores of next tokens from previous beams. Since this is a summation for all the logits - topK should not be affected by this? Doing this after topK is more efficient?

Reply:
Topk is on each batch. The added scores is per beam so this shall be done before top-k otherwise the result will be different.